usdigitalresponse/grants-ingest
# Enhancement: #66 upload Lambda artifacts only when changed

### Resolves #66 

## Description

This PR introduces a new, reusable/helper Terraform module at `terraform/modules/taskfile_lambda_builder`. As the name implies, it uses commands defined in `Taskfile.yml` to build Lambda function artifacts. It differs from the existing solution for building artifacts via the [`terraform-aws-modules/lambda/aws` module](https://github.com/terraform-aws-modules/terraform-aws-lambda) in that it compiles binaries (or more accurately, runs the Taskfile commands to do so) and zip file artifacts during `terraform plan`. By contrast, the existing solution waits until `terraform apply` to complete these steps, presumably in an effort to keep `terraform plan` somewhat lightweight and wait until `apply` to actually execute. However, the nature of the existing solution means that Lambda function artifacts are rebuilt on every execution of `terraform apply`, rather than only when an artifact has actually changed. This makes plan output less useful due to the presence of "perma-diffs" and needlessly expends effort uploading largely-unchanged artifacts to S3.

Note: In order to generate a checksum for comparison, a local build still needs to exist. If it doesn't, it will be created automatically. However, by leveraging the Taskfile system's features for [avoiding unnecessary work](https://taskfile.dev/usage/#prevent-unnecessary-work), binaries are only rebuilt locally on Taskfile cache miss (when source code is modified, or when the Taskfile cache is empty). Furthermore, regardless of whether a Taskfile command results in a new binary, an artifact will only be uploaded to S3 when its MD5 checksum differs from the corresponding remote artifact.

TL;DR the new module and accompanying refactored Terraform code enables us to upload a new Lambda artifact only if it differs from what already exists on S3 (or LocalStack).

## Testing

Optional: Before testing and/or between each step, ensure that `bin/` and `terraform/builds` directories are empty. Although binaries and zip artifacts will be created if/when missing, they should have no impact on the test results.

1. Run `terraform apply`, targeting a LocalStack environment.
2. Run `terraform apply` again (or simply, `terraform plan`). Verify that no changes were detected by Terraform.
3. Modify the source code for a Lambda function in some way (e.g. by adding debug log output).
4. Run `terraform plan`. Verify that Terraform only detects changes for the modified Lambda function.
5. Undo the change from Step 3.
6. Run `terraform apply` or `terraform plan`. Verify that no changes were detected by Terraform.

### Automated and Unit Tests

### Manual tests for Reviewer

## Checklist
